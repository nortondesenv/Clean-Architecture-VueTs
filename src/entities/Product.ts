export interface IProduct {
    available: boolean;
    image: string;
    name: string;
    priceMember: number;
    pricePromotional: number;
    priceStock: number;
}
