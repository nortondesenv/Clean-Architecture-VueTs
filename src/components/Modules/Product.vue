<template>
    <div class="Card__Product">
        <div class="Card__Product__Container">
            <div class="Card__Product__Container__Image">
                <img :src="product.image" />
            </div>

            <div class="Card__Product__Container__Info">
                <div class="Card__Product__Container__Info__Text">
                    {{ product.name }}
                </div>
                <div
                    v-if="product.available"
                    class="Card__Product__Container__Info__Text--Promotional"
                >
                    {{ pricePromotionalFormated }}
                </div>
                <div
                    v-if="product.available"
                    class="Card__Product__Container__Info__Text--Socio"
                >
                    SOCIO WINE
                </div>
                <div
                    v-if="product.available"
                    class="Card__Product__Container__Info__Text--Member"
                >
                    {{ priceMemberFormated }}
                </div>

                <Button
                    :autoWidth="true"
                    :disabled="!product.available"
                    margin="Margin-Small"
                    type="Secundary"
                    :text="buttonText"
                    @click="addProduct()"
                ></Button>
            </div>
        </div>
    </div>
</template>

<script lang="ts">
import Vue from 'vue';
import Button from '@/components/Base/Button.vue';
import AddItemToCartUseCase, {
    IAddItemToCartUseCase,
} from '@/usecases/AddItemToCartUseCase';
import { IProduct } from '@/entities/Product';
import errorService from '@/services/ErrorService';
import { CartEntity } from '@/entities/Cart';

export default Vue.extend({
    props: {
        product: {
            type: Object as () => IProduct,
            required: true,
        },
    },
    methods: {
        addProduct() {
            const params: IAddItemToCartUseCase = {
                cart: new CartEntity(),
                errorService: new errorService({
                    context: 'Adicionar o produto no carrinho.',
                }),
            };
            new AddItemToCartUseCase(params).execute(this.product);
        },
    },
    components: {
        Button,
    },
    computed: {
        buttonText(): String {
            return this.product.available ? 'Adicionar' : 'Esgotado';
        },
        pricePromotionalFormated(): Number {
            return this.product.pricePromotional.toLocaleString('pt-br', {
                style: 'currency',
                currency: 'BRL',
            });
        },
        priceMemberFormated(): Number {
            return this.product.priceMember.toLocaleString('pt-br', {
                style: 'currency',
                currency: 'BRL',
            });
        },
    },
});
</script>
