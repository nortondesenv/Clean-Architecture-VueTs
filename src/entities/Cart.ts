import { IProduct } from '@/entities/Product';
import store from '@/store';
import { AddItem, ClearItems, ToggleCart } from '@/store/modules/cart/types';

export interface ICartItem extends IProduct {
    checkoutPrice: number;
    amount: number;
}

export interface ICart {
    items: ICartItem[];
}

export class CartEntity {
    constructor() {}

    async toggleCart() {
        await store.commit(new ToggleCart(store.state.cart.isCartOpen));
    }

    removeItem(index: number) {
        store.state.cart.cart.items.splice(index, 1);
    }

    async addItemToCart(item: IProduct) {
        const checkoutPrice = item.priceMember;
        const amount = 1;
        const cartItem = {
            ...item,
            checkoutPrice,
            amount,
        };
        await store.commit(new AddItem(cartItem));
        alert('Produto adicionado em seu carrinho.');
    }

    async checkOut() {
        // Start operation
        store.commit(new ClearItems());
        alert('Obrigado e volte sempre!');
        return;
    }
}
