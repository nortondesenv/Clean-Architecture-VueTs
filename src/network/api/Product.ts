import { AxiosResponse } from 'axios';
import { APIRequest } from '@/network/APIRequest';
import { HTTPMethod } from '@/network/APIClient';
import { IProduct } from '@/entities/Product';

export namespace ProductAPI {
    // List products
    export class GetProducts implements APIRequest<List<IProduct>> {
        response: List<IProduct>;
        path = '/908ec5b5-1e5a-4602-9008-47719f7c6759';
        method = HTTPMethod.GET;
        parse = (data: AxiosResponse) => data.data;
        constructor(public params: null) {}
    }
}
