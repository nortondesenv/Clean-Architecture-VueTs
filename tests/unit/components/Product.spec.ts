import { shallowMount } from '@vue/test-utils';
import Product from '@/components/modules/Product.vue';
import { IProduct } from '@/entities/Product';

describe('Product.vue', () => {
    const product: IProduct = {
        available: true,
        image: 'testeImage.png',
        name: 'testProduct',
        priceMember: 5.55,
        pricePromotional: 5.55,
        priceStock: 5.55,
    };

    it('Renderiza Product disponivel', () => {
        const wrapper = shallowMount(Product, {
            propsData: { product: product },
        });
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Renderiza Product nÃ£o disponivel', () => {
        product.available = false;
        const wrapper = shallowMount(Product, {
            propsData: { product: product },
        });
        expect(wrapper.html()).toMatchSnapshot();
    });
});
