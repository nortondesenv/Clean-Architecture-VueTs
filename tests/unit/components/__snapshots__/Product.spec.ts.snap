// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product.vue Renderiza Product disponivel 1`] = `
<div class="Card__Product">
  <div class="Card__Product__Container">
    <div class="Card__Product__Container__Image"><img src="testeImage.png"></div>
    <div class="Card__Product__Container__Info">
      <div class="Card__Product__Container__Info__Text">
        testProduct
      </div>
      <div class="Card__Product__Container__Info__Text--Promotional">
        R$&nbsp;5,55
      </div>
      <div class="Card__Product__Container__Info__Text--Socio">
        SOCIO WINE
      </div>
      <div class="Card__Product__Container__Info__Text--Member">
        R$&nbsp;5,55
      </div>
      <button-stub nativetype="button" type="Secundary" size="Midium" autowidth="true" margin="Margin-Small" text="Adicionar"></button-stub>
    </div>
  </div>
</div>
`;

exports[`Product.vue Renderiza Product n√£o disponivel 1`] = `
<div class="Card__Product">
  <div class="Card__Product__Container">
    <div class="Card__Product__Container__Image"><img src="testeImage.png"></div>
    <div class="Card__Product__Container__Info">
      <div class="Card__Product__Container__Info__Text">
        testProduct
      </div>
      <!---->
      <!---->
      <!---->
      <button-stub nativetype="button" disabled="true" type="Secundary" size="Midium" autowidth="true" margin="Margin-Small" text="Esgotado"></button-stub>
    </div>
  </div>
</div>
`;
